server/src/unifyfs_tree.c

```diff
@@ -476,3 +476,318 @@ static void filesize_response_rpc(hg_handle_t handle)
+    assert(hret == HG_SUCCESS);
+
+    /* decode response */
+    rc = (int) out.ret;
+
+    /* free resources */
+    margo_free_output(handle, &out);
+    margo_destroy(handle);
+
+    return rc;
+}
+
+static int rpc_invoke_truncate_response(int rank, unifyfs_state_truncate_t* st)
+{
+    int rc = (int)UNIFYFS_SUCCESS;
+
+    /* get address for specified server rank */
+    hg_addr_t addr = glb_servers[rank].margo_svr_addr;
+
+    /* get handle to rpc function */
+    hg_handle_t handle;
+    hg_return_t hret = margo_create(unifyfsd_rpc_context->svr_mid, addr,
+        unifyfsd_rpc_context->rpcs.truncate_response_id, &handle);
+    assert(hret == HG_SUCCESS);
+
+    /* fill in input struct */
+    truncate_response_in_t in;
+    in.tag      = (int32_t) st->parent_tag;
+    in.err      = (int32_t) st->err;
+
+    /* call rpc function */
+    hret = margo_forward(handle, &in);
+    assert(hret == HG_SUCCESS);
+
+    /* wait for rpc output */
+    truncate_response_out_t out;
+    hret = margo_get_output(handle, &out);
+    assert(hret == HG_SUCCESS);
+
+    /* decode response */
+    rc = (int) out.ret;
+
+    /* free resources */
+    margo_free_output(handle, &out);
+    margo_destroy(handle);
+
+    return rc;
+}
+
+unifyfs_state_truncate_t* state_truncate_alloc(
+    int root,
+    int gfid,
+    size_t length,
+    int32_t ptag,
+    int32_t tag)
+{
+    unifyfs_state_truncate_t* st = (unifyfs_state_truncate_t*)
+        malloc(sizeof(unifyfs_state_truncate_t));
+    st->root          = root;
+    st->gfid          = gfid;
+    st->parent_tag    = ptag;
+    st->tag           = tag;
+    st->num_responses = 0;
+    st->length        = length;
+    st->err           = UNIFYFS_SUCCESS;
+    st->mutex         = ABT_MUTEX_NULL;
+    st->cond          = ABT_COND_NULL;
+
+    /* define tree to use for given root */
+    unifyfs_tree_init(glb_pmi_rank, glb_pmi_size, root, 2,
+        &st->tree);
+
+    return st;
+}
+
+void state_truncate_free(unifyfs_state_truncate_t** pst)
+{
+    if (pst != NULL) {
+        unifyfs_state_truncate_t* st = *pst;
+
+        unifyfs_tree_free(&st->tree);
+
+        if (st->cond != ABT_COND_NULL) {
+            ABT_cond_free(&st->cond);
+        }
+        if (st->mutex != ABT_MUTEX_NULL) {
+            ABT_mutex_free(&st->mutex);
+        }
+
+        free(st);
+
+        *pst = NULL;
+    }
+}
+
+static void truncate_response_forward(unifyfs_state_truncate_t* st)
+{
+    printf("%d: BUCKEYES response_forward (truncate)\n", glb_pmi_rank);
+    fflush(stdout);
+    /* get tree we're using for this operation */
+    unifyfs_tree_t* t = &st->tree;
+
+    /* get info for tree */
+    int parent       = t->parent_rank;
+    int child_count  = t->child_count;
+    int* child_ranks = t->child_ranks;
+
+    /* send up to parent if we have gotten all replies */
+    if (st->num_responses == child_count) {
+        /* truncate the file */
+        int ret = gfid2ext_tree_truncate(&glb_gfid2ext, st->gfid, st->length);
+
```

Comment:
Looks great!  The only thing I can think we might add is to set the st->err flag here if the truncate operation were to fail.